rfcs/quic.md

```diff
@@ -0,0 +1,104 @@
+[aioquic](https://github.com/aiortc/aioquic) in Python 3 and listening on port
+`9001` by default. This server is used to test web APIs that rely on the
+QUIC protocol (e.g. [WebTransport](https://wicg.github.io/web-transport/)). The
+server is optional but tries to start by default; `wpt` front-end will skip the
+tests that rely on QUIC if the QUIC server cannot be started (e.g. due to
+missing dependencies).
+
+QUIC custom handlers use a new filename flag `.quic` (`.quic.py`). And a meta
+tag is included by tests that require the QUIC server.
+
+See the [original issue][original-issue] for more background.
+
+## Details
+
+### Requirements
+
+1.  The QUIC server requires decent cryptography support; pure Python isn't
+    sufficient.
+2.  The server needs to be scriptable, and the custom handlers should live in
+    the wpt repo next to test files, so that they can be updated together.
+3.  An async programming model is preferred.
+
+### Architecture
+
+The QUIC server has a standalone CLI, and is managed by `wpt serve` as a
+subprocess. Configurations are stored in `config.json` (same as `wptserve`), and
+`wpt serve` is responsible for passing the relevant configurations (e.g. port
+numbers) to the QUIC server via command line flags. `wpt serve` handles the
+error gracefully if the QUIC server cannot be started.
+
+The QUIC server is written in Python 3 and makes no attempt to support Python 2.
+Its entrypoint will have a shebang that explicitly requires `python3`. An
+implication is that the QUIC server will be relatively self contained, without
+dependencies on other modules in wpt that are currently Python 2-only.
+
+Code for the QUIC server itself will live in `tools/quic`.
+
+### User interface
+
+For **test writers**: the QUIC server will support Python custom handlers
+similar to `wptserve`. The APIs will likely be different. To more easily
+distinguish the two, we add a filename flag to QUIC custom handlers
+(`.quic.py`). These handlers will be ignored by `wptserve`. Tests that require
+the QUIC server need to be marked with a meta tag (tentatively `<meta
+name="quic" content="true">` and `// META: quic=true`). This information will be
+extracted by `wpt manifest` into `MANIFEST.json`, so that `wptrunner` can skip
+the tests that require the QUIC server easily.
+
+For **test runners**: the goal is to not introduce any observable changes to
+infrastructure maintainers not yet concerned with QUIC or WebTransport. All
+existing tests will continue to run without any infra changes, and all new
+QUIC-dependent tests will be skipped unless all dependencies are met (or can
+also be skipped explicitly).
+
+### Notable dependencies
+
+* [sys] Python 3 (>=3.6)
+* [py] aioquic
+    * [py] cryptography
+        * [sys] openssl (>=1.0.1, already required by [HTTP/2](http2.md))
+    * [py] pylsqpack
+
+All of the Python dependencies above have native extensions, which requires
```

Comment:
So it seems like updating the version here is going to be an infra-breaking change for vendors because work will be required on the vendor side to update to the new version (e.g. updating the deps file or uploading the wheel to the internal pypi mirror or whatever each vendor does). That seems like a pretty big concern to me. If we did vendor the source, we could have logic like "check if the correct version is available, or pip install, or build from source". Maybe there are other possibilities to address this?

Reply:
I'm thinking about this from two perspectives:

* For average human users, it's most likely they are using a platform supported by PyPI wheels, and the dependencies will be installed into `_venv` using `pip` automatically. Vendoring the source isn't very useful for them.
* For browser infras, AFAIK both Firefox and Chromium discourage or even disallow building a native Python extension at run time; these dependencies will have to come from a PyPI-like repo. If that's the case, vendoring the source doesn't seem helpful, either.

I agree that this will incur a maintenance burden on the vendors if they want to run QUIC tests. (I'll acknowledge this in the risks section.) Meanwhile, I don't see vendoring the source code would help the browser vendors if my understanding above is correct -- someone would still need to build and distribute the new version.

Reply:
I don't think our test environment forbids building a native extension, although it also seems likely that the required libraries aren't available. What would definitely be possible (although not tivial) is adding the build step to the Firefox build and distributing the resulting package to the test machines; we already have various test-only binaries that need to be produced.

What worries me about the proposed setup is that updating the library version is a sync-blocking change for Gecko since we're a) using the tools/ directory directly and b) don't allow PyPI access from test machines for robustness reasons.

Reply:
That's true for Chromium, too. I imagine that in the future when we need to update these native dependencies, we will need to create a PR that updates the `requirements.txt` for venv, and ask each vendor to prepare the packages in their infras, before merging the PR. Native dependencies are difficult to deal with in general; I'm intentionally making some tradeoff here between ongoing maintenance costs and complexity.

Do you have a more concrete alternative proposal? If you think vendoring the source into the codebase is easier for Firefox infra, then I'd be happy to add that as a fallback to wheels, too. I just didn't want to vendor them if it wouldn't be used by anyone.

Reply:
I think that adding the source would *in principle* be better for Mozilla, since we could in theory add it to the build system. In practice I don't know that we'd actually get around to doing that :/ It probably depends on how hard it is vs doing manual updates. But I think I'd like us to do it anyway.

In any case I think the process probably demands an RFC for version updates, since it's a change that affects multiple vendors. That's pretty heavyweight, but this really is the kind of change that can break integrations.